// vuex-webextensions/dist/index.js
"use strict";
Object.defineProperty(exports, "__esModule", {
    value: !0
}), exports["default"] = _default;
var _backgroundScript = _interopRequireDefault(require("./backgroundScript")),
    _browser = _interopRequireDefault(require("./browser")),
    _contentScript = _interopRequireDefault(require("./contentScript")),
    _logger = _interopRequireDefault(require("./logger"));

function _interopRequireDefault(a) {
    return a && a.__esModule ? a : {
        default: a
    }
}

function ownKeys(a, b) {
    var c = Object.keys(a);
    if (Object.getOwnPropertySymbols) {
        var d = Object.getOwnPropertySymbols(a);
        b && (d = d.filter(function(b) {
            return Object.getOwnPropertyDescriptor(a, b).enumerable
        })), c.push.apply(c, d)
    }
    return c
}

function _objectSpread(a) {
    for (var b, c = 1; c < arguments.length; c++) b = null == arguments[c] ? {} : arguments[c], c % 2 ? ownKeys(b, !0).forEach(function(c) {
        _defineProperty(a, c, b[c])
    }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(a, Object.getOwnPropertyDescriptors(b)) : ownKeys(b).forEach(function(c) {
        Object.defineProperty(a, c, Object.getOwnPropertyDescriptor(b, c))
    });
    return a
}

function _defineProperty(a, b, c) {
    return b in a ? Object.defineProperty(a, b, {
        value: c,
        enumerable: !0,
        configurable: !0,
        writable: !0
    }) : a[b] = c, a
}
var defaultOptions = {
    connectionName: "vuex-webextensions",
    loggerLevel: "warning",
    persistentStates: [],
    ignoredMutations: [],
    ignoredActions: [],
    syncActions: !0
};

function _default(a) {
    if ("undefined" == typeof window) return function() {};
    var b = _objectSpread({}, defaultOptions, {}, a);
    _logger["default"].setLoggerLevel(b.loggerLevel);
    var c = new _browser["default"];
    return function(a) {
        a.registerModule("@@VWE_Helper", {
            mutations: {
                vweReplaceState: function vweReplaceState(b, c) {
                    Object.keys(a.state).forEach(function(b) {
                        a.state[b] = c[b]
                    })
                }
            }
        }), c.isBackgroundScript(window).then(function(d) {
            return d ? new _backgroundScript["default"](a, c, b) : new _contentScript["default"](a, c, b)
        })
    }
}
